MySQL database settings stopped working on upgrade to v0.60.3.8

### Describe the bug

I was testing upgrading from 0.59 to 0.60 with a clone of the app db from my production instance. The test MySQL database connection in that instance stopped working in a bizarre way. At first it was due to missing permissions, which I fixed. I can confirm I can connect with the mysql command line tool from the same host, but Metabase is still having a problem. It seems to be looking for a particular response from the MySQL server, but it isn't getting what it wants?

Once this happens a few times, something gets corrupted as the exception changes to an NPE with no traceback.

~Creating a new connection with the same parameters works and connects successfully.~ (Not correct)

Packet dump shows that Metabase isn't getting to the point of opening the connection to the database, so something seems messed up in the app db.


### To Reproduce

This was an upgrade from v0.59.9.1 to v0.60.3.8. The mysql connection has been there since February so it probably predates 0.59.

### Expected behavior

_No response_

### Logs

The first set of exceptions:
```
2026-05-06 18:16:45,848 ERROR driver.util :: Failed to connect to Database 
java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
	at java.base/java.util.regex.Matcher.getTextLength(Matcher.java:1808)
	at java.base/java.util.regex.Matcher.reset(Matcher.java:461)
	at java.base/java.util.regex.Matcher.<init>(Matcher.java:256)
	at java.base/java.util.regex.Pattern.matcher(Pattern.java:1180)
	at clojure.core$re_matcher.invokeStatic(core.clj:4909)
	at clojure.core$re_find.invokeStatic(core.clj:4951)
	at clojure.core$re_find.invoke(core.clj:4951)
	at metabase.driver.mysql$fn__164273.invokeStatic(mysql.clj:219)
	at metabase.driver.mysql$fn__164273.invoke(mysql.clj:215)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__64056.invoke(util.clj:136)
	at clojure.core$binding_conveyor_fn$fn__5844.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
2026-05-06 18:16:45,848 ERROR warehouses.util :: Cannot connect to Database 
clojure.lang.ExceptionInfo: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null {:message "Cannot invoke \"java.lang.CharSequence.length()\" because \"this.text\" is null"}
	at metabase.driver.util$can_connect_with_details_QMARK_.invokeStatic(util.clj:153)
	at metabase.driver.util$can_connect_with_details_QMARK_.doInvoke(util.clj:124)
	at clojure.lang.RestFn.invoke(RestFn.java:445)
	at metabase.warehouses.util$test_database_connection.invokeStatic(util.clj:44)
	at metabase.warehouses.util$test_database_connection.doInvoke(util.clj:34)
	at clojure.lang.RestFn.invoke(RestFn.java:428)
	at metabase.warehouses_rest.api$fn__143621$fn__143628.invoke(api.clj:1170)
	at metabase.api.macros$_core_fn$core_fn__94419.invoke(macros.clj:427)
	at metabase.api.macros$endpoint_handler_STAR_$handler__94468.invoke(macros.clj:635)
	at metabase.api.macros.scope$ensure_scopes_checked$fn__93573.invoke(scope.clj:79)
	at metabase.api.macros$build_ns_handler$ns_handler_STAR___94494.invoke(macros.clj:709)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.routes.common$enforce_authentication$fn__94956.invoke(common.clj:92)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__94828.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at metabase.api_routes.routes$fn__159445.invokeStatic(routes.clj:124)
	at metabase.api_routes.routes$fn__159445.invoke(routes.clj:123)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765.invoke(core.clj:200)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at metabase.server.routes$api_handler$api_handler_STAR___119316.invoke(routes.clj:93)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765.invoke(core.clj:200)
	at compojure.core$make_context$handler__94793.invoke(core.clj:290)
	at compojure.core$make_context$fn__94797.invoke(core.clj:300)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__94746.invoke(core.clj:153)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__94746.invoke(core.clj:153)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__94746.invoke(core.clj:153)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__94746.invoke(core.clj:153)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__94746.invoke(core.clj:153)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__94746.invoke(core.clj:153)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__94746.invoke(core.clj:153)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$make_context$fn__94797.invoke(core.clj:301)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$make_context$fn__94797.invoke(core.clj:301)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:199)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__94828.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__94828.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765$f__94766$respond_SINGLEQUOTE___94767.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__94828.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765.invoke(core.clj:200)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at compojure.core$routes$fn__94765$f__94766.invoke(core.clj:198)
	at compojure.core$routes$fn__94765.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__121254.invoke(exceptions.clj:87)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__121251.invoke(exceptions.clj:76)
	at metabase.server.middleware.log$log_api_call$fn__120735$fn__120736$fn__120737.invoke(log.clj:241)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invokeStatic(diagnostic.clj:17)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invoke(diagnostic.clj:12)
	at metabase.server.middleware.log$log_api_call$fn__120735$fn__120736.invoke(log.clj:232)
	at toucan2.execute$do_with_call_counts.invokeStatic(execute.clj:112)
	at toucan2.execute$do_with_call_counts.invoke(execute.clj:103)
	at metabase.server.middleware.log$log_api_call$fn__120735.invoke(log.clj:231)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__121518.invoke(browser_cookie.clj:40)
	at metabase.server.middleware.security$add_security_headers$fn__121217.invoke(security.clj:335)
	at metabase.server.middleware.json$wrap_json_body$fn__120844.invoke(json.clj:128)
	at metabase.server.middleware.offset_paging$handle_paging$fn__121342.invoke(offset_paging.clj:48)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__120832.invoke(json.clj:83)
	at metabase.server.middleware.metadata_provider_cache$wrap_metadata_provider_cache$fn__121528.invoke(metadata_provider_cache.clj:11)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__121793.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__121812.invoke(params.clj:77)
	at metabase.server.middleware.auth$verify_slack_request$fn__120762.invoke(auth.clj:86)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__120380.invoke(misc.clj:78)
	at metabase.server.middleware.session$reset_session_timeout$fn__121505.invoke(session.clj:326)
	at metabase.server.middleware.session$bind_current_user$fn__121488$fn__121489.invoke(session.clj:283)
	at metabase.request.session$do_with_current_user$fn__88577.invoke(session.clj:59)
	at metabase.request.session$do_with_current_user$do_with_user_local_values__88572.invoke(session.clj:55)
	at metabase.request.session$do_with_current_user.invokeStatic(session.clj:56)
	at metabase.request.session$do_with_current_user.invoke(session.clj:36)
	at metabase.server.middleware.session$bind_current_user$fn__121488.invoke(session.clj:282)
	at metabase.server.middleware.session$wrap_current_user_info$fn__121472.invoke(session.clj:258)
	at metabase.server.middleware.premium_features_cache$wrap_premium_features_cache_check$fn__121320.invoke(premium_features_cache.clj:57)
	at metabase.server.middleware.settings_cache$wrap_settings_cache_check$fn__121287.invoke(settings_cache.clj:50)
	at metabase.analytics.sdk$embedding_mw$embedding_mw_fn__63080.invoke(sdk.clj:69)
	at metabase.server.middleware.session$wrap_session_key$fn__121416.invoke(session.clj:91)
	at metabase.server.middleware.auth$wrap_static_api_key$fn__120754.invoke(auth.clj:48)
	at ring.middleware.cookies$wrap_cookies$fn__121713.invoke(cookies.clj:200)
	at metabase.server.middleware.misc$add_version$fn__120342.invoke(misc.clj:29)
	at metabase.server.middleware.misc$add_content_type$fn__120350.invoke(misc.clj:47)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__120388.invoke(misc.clj:94)
	at ring.middleware.gzip$wrap_gzip$fn__121755.invoke(gzip.clj:86)
	at metabase.server.middleware.trace$wrap_trace$fn__120861.invoke(trace.clj:116)
	at metabase.server.middleware.request_id$wrap_request_id$fn__119353.invoke(request_id.clj:10)
	at metabase.server.middleware.misc$bind_request$fn__120391$fn__120392.invoke(misc.clj:106)
	at metabase.request.current$do_with_current_request.invokeStatic(current.clj:19)
	at metabase.request.current$do_with_current_request.invoke(current.clj:15)
	at metabase.server.middleware.misc$bind_request$fn__120391.invoke(misc.clj:105)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__121542.invoke(ssl.clj:42)
	at metabase.server.instance$async_proxy_handler$fn__120423.invoke(instance.clj:89)
	at metabase.server.instance.proxy$org.eclipse.jetty.ee9.servlet.ServletHandler$ff19274a.doHandle(Unknown Source)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.handle(ScopedHandler.java:125)
	at metabase.server.statistics_handler$new_handler$fn__112435.invoke(statistics_handler.clj:106)
	at metabase.server.statistics_handler.proxy$org.eclipse.jetty.ee9.nested.HandlerWrapper$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.ee9.nested.HandlerWrapper.handle(HandlerWrapper.java:124)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextHandle(ScopedHandler.java:195)
	at org.eclipse.jetty.ee9.nested.ContextHandler.doHandle(ContextHandler.java:1047)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextScope(ScopedHandler.java:162)
	at org.eclipse.jetty.ee9.servlet.ServletHandler.doScope(ServletHandler.java:483)
	at metabase.server.instance.proxy$org.eclipse.jetty.ee9.servlet.ServletHandler$ff19274a.doScope(Unknown Source)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextScope(ScopedHandler.java:160)
	at org.eclipse.jetty.ee9.nested.ContextHandler.doScope(ContextHandler.java:962)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.handle(ScopedHandler.java:123)
	at org.eclipse.jetty.ee9.nested.ContextHandler.handle(ContextHandler.java:1729)
	at org.eclipse.jetty.ee9.nested.HttpChannel$RequestDispatchable.dispatch(HttpChannel.java:1581)
	at org.eclipse.jetty.ee9.nested.HttpChannel.dispatch(HttpChannel.java:733)
	at org.eclipse.jetty.ee9.nested.HttpChannel.handle(HttpChannel.java:520)
	at org.eclipse.jetty.ee9.nested.ContextHandler$CoreContextHandler$CoreToNestedHandler.handle(ContextHandler.java:3054)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1224)
	at org.eclipse.jetty.server.Server.handle(Server.java:197)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:787)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:420)
	at org.eclipse.jetty.server.internal.HttpConnection$FillableCallback.succeeded(HttpConnection.java:1809)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:54)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:1009)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1239)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1194)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
	at java.base/java.util.regex.Matcher.getTextLength(Matcher.java:1808)
	at java.base/java.util.regex.Matcher.reset(Matcher.java:461)
	at java.base/java.util.regex.Matcher.<init>(Matcher.java:256)
	at java.base/java.util.regex.Pattern.matcher(Pattern.java:1180)
	at clojure.core$re_matcher.invokeStatic(core.clj:4909)
	at clojure.core$re_find.invokeStatic(core.clj:4951)
	at clojure.core$re_find.invoke(core.clj:4951)
	at metabase.driver.mysql$fn__164273.invokeStatic(mysql.clj:219)
	at metabase.driver.mysql$fn__164273.invoke(mysql.clj:215)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__64056.invoke(util.clj:136)
	at clojure.core$binding_conveyor_fn$fn__5844.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	... 1 more
2026-05-06 18:16:45,849 DEBUG middleware.log :: PUT /api/database/23 400 4ms (1 DB calls) {:metabase-user-id 1} 
{:message "Cannot invoke \"java.lang.CharSequence.length()\" because \"this.text\" is null"}
```
It eventually becomes this:
```
2026-05-06 18:23:04,486 ERROR driver.util :: Failed to connect to Database 
java.lang.NullPointerException
2026-05-06 18:23:04,487 ERROR warehouses.util :: Cannot connect to Database 
java.lang.NullPointerException
2026-05-06 18:23:04,487 DEBUG middleware.log :: PUT /api/database/23 400 3ms (1 DB calls) {:metabase-user-id 1} 
{:message nil}
```

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:150.0) Gecko/20100101 Firefox/150.0",
    "vendor": ""
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "h2",
      "mysql"
    ],
    "run-mode": "prod",
    "plan-alias": "",
    "version": {
      "date": "2026-05-04",
      "tag": "v0.60.3.8",
      "hash": "95ddde1"
    },
    "settings": {
      "report-timezone": null
    },
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "18.3"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.8"
      }
    }
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "21.0.10+7-1",
    "java.vendor": "OpenJDK BSD Porting Team",
    "java.vendor.url": "https://github.com/battleblow/jdk21u/",
    "java.version": "21.0.10",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "21.0.10+7-1",
    "os.name": "FreeBSD",
    "os.version": "14.4-RELEASE-p3",
    "user.language": "en",
    "user.timezone": "America/Los_Angeles"
  }
}
```

### Severity

Not a major problem, but curious

### Additional context

I can safely delete/recreate this database in Metabase without losing anything, so this is more a curiosity in case someone else runs into it. But still, it shouldn't happen.

I decrypted the connection details for the troublesome db and it looks like this. Host and password have been sanitized. `"additional-options":null` looks problematic.

```
{"role":null,"additional-options":null,"ssl":true,"password":"notreallythepassword","let-user-control-scheduling":true,"destination-database":false,"port":3306,"advanced-options":true,"dbname":"test","ssl-cert":null,"host":"dbserver.example.com","tunnel-enabled":false,"json-unfolding":true,"user":"metabase"}
```

--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -218,7 +218,7 @@
   [driver details]
   ;; delegate to parent method to check whether we can connect; if so, check if it's an unsupported version and issue
   ;; a warning if it is
-  (let [match (re-find disallowed-additional-opts (:additional-options details ""))]
+  (let [match (some->> (:additional-options details) (re-find disallowed-additional-opts))]
     (when match
       (throw (ex-info "Potentially dangerous keys in additional options" {:disallowed-key match}))))
   (when ((get-method driver/can-connect? :sql-jdbc) driver details)

--- a/test/metabase/driver/mysql_test.clj
+++ b/test/metabase/driver/mysql_test.clj
@@ -1042,21 +1042,26 @@
               (jdbc/execute! spec "DROP USER IF EXISTS 'partial_revokes_test_user';"))))))))
 
 (deftest ^:parallel only-connect-when-non-malicious-properties
-  (testing "Reject connection strings with malicious properties"
-    (are [bad-option] (let [details (assoc (tx/dbdef->connection-details :mysql :db
-                                                                         {:database-name "argent"})
-                                           :additional-options bad-option)
-                            result (try (driver/can-connect? :mysql details)
-                                        ::did-not-throw
-                                        (catch Exception e e))]
-                        (is (instance? clojure.lang.ExceptionInfo result))
-                        (is (partial= {:cause "Potentially dangerous keys in additional options"}
-                                      (Throwable->map result))))
-      "allowLoadLocalInfile=true"
-      "allowLoadLocalInfileInPath=1"
-      "allowUrlInLocalInfile=1"
-      "autoDeserialize=1"
-      "serverRSAPublicKeyFile=/path/to/file")))
+  (mt/test-driver :mysql
+    (let [details (:details (mt/db))]
+      (testing "Reject connection strings with malicious properties"
+        (are [bad-option] (let [details (assoc details :additional-options bad-option)]
+                            (is (thrown-with-msg? clojure.lang.ExceptionInfo
+                                                  #"Potentially dangerous keys in additional options"
+                                                  (driver/can-connect? :mysql details))))
+          "allowLoadLocalInfile=true"
+          "allowLoadLocalInfileInPath=1"
+          "allowUrlInLocalInfile=1"
+          "autoDeserialize=1"
+          "serverRSAPublicKeyFile=/path/to/file"))
+      (testing "Allow connection strings with non-malicious properties"
+        (are [ok-option] (let [details (assoc details :additional-options ok-option)]
+                           (is (true? (driver/can-connect? :mysql details))))
+          nil
+          ""
+          " "
+          "tinyInt1isBit=1")
+        (is (true? (driver/can-connect? :mysql details)))))))
 
 (deftest ^:parallel set-role-statement-escape-quotes-test
   (mt/test-driver :mysql

